File: src/javaboost/weaklearning/SingleFeatureSigmoidClassifier.java
package javaboost.weaklearning;

public class SingleFeatureSigmoidClassifier implements WeakClassifier{

    static final long serialVersionUID = -4227240934866508505L;

    private int m_featColumn;
    private float m_threshold;
    private double m_leftConf = 0;
    private double m_rightConf = 0;
    private double m_smoothW = 0;
    private double m_dcBias = 0;
    public SingleFeatureSigmoidClassifier(int featColumn,
					  float threshold,
					  double smoothW,
					  double leftConf,
					  double rightConf,
					  double dcBias) {
	m_featColumn = featColumn;
	m_threshold = threshold;
	m_leftConf= leftConf;
	m_rightConf = rightConf;
	m_smoothW = smoothW;
	m_dcBias = dcBias;
    }

    public double[] classify(float[][] data){
	double[] output = new double[data.length];
	double alpha = m_rightConf - m_leftConf;
	double bias = m_leftConf;
	for(int i = 0; i < data.length; ++i) {
	    if(Float.isInfinite(data[i][m_featColumn])) {
		output[i] = m_dcBias;
	    }
	    else{
		output[i] = bias + alpha/(1+Math.exp(-m_smoothW*(data[i][m_featColumn]-m_threshold)));
	    }

	}
	return output;
    }

    public String toString() {
	return new String("Sigmoid: "
			  +m_featColumn
			  +"@"
			  +m_threshold
			  +" SmoothingWT:"
			  +m_smoothW
			  +" less: "
			  +m_leftConf
			  +" greaterOrEqual: "
			  +m_rightConf
			  +" DC: "
			  +m_dcBias);
    }
    public double getLB() {
	return Math.min(m_dcBias, Math.min(m_leftConf, m_rightConf));
    }

    public double getUB() {
	return Math.max(m_dcBias, Math.max(m_leftConf, m_rightConf));
    }


    public int[] getColumns() {
	int[] out = {m_featColumn};
	return out;
    }

}

File: C++/src/singlefeaturesigmoidclassifier.cpp
#include "singlefeaturesigmoidclassifier.h"
#include "utils.h"
#include <cmath>

using std::vector;

SingleFeatureSigmoidClassifier::
SingleFeatureSigmoidClassifier(int featColumn,
                               float threshold,
                               float smoothW,
                               float lessConf,
                               float grtrConf,
                               float dcBias) : m_featColumn(featColumn), m_threshold(threshold),
    m_lessConf(lessConf), m_grtrConf(grtrConf), m_smoothW(smoothW), m_dcBias(dcBias) {}


void SingleFeatureSigmoidClassifier::
classify(float* output, float** data, int N, int NC) {
    SingleFeatureSigmoidClassifier::classify(output, data, N, NC, m_lessConf, m_grtrConf, m_dcBias, m_smoothW, m_threshold, m_featColumn);

}

void SingleFeatureSigmoidClassifier::
classify(float* output, float ** data, int N, int NC, float lessConf, float grtrConf, float dcBias, float smoothingW, float threshold, int featColumn) {
    float alpha = grtrConf - lessConf;
    float bias = lessConf;
    for(int i = 0; i < N; ++i) {
        if(utils::isinf(data[i][featColumn])) {
            output[i] = dcBias;
        } else{
            output[i] = bias + alpha/(1+exp(-smoothingW*(data[i][featColumn]-threshold)));
        }
    }
}

vector< vector<float> > SingleFeatureSigmoidClassifier::
getParams() {
   vector< vector< float > > all_params;
   vector<float> params;
   params.push_back((float)m_featColumn);
   params.push_back(m_threshold);
   params.push_back(m_smoothW);
   params.push_back(m_lessConf);
   params.push_back(m_grtrConf);
   params.push_back(m_dcBias);
   all_params.push_back(params);
   return all_params;
}
